src/main/java/org/opensearch/neuralsearch/highlight/extractor/QueryTextExtractorRegistry.java

```diff
@@ -0,0 +1,90 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+package org.opensearch.neuralsearch.highlight.extractor;
+
+import org.apache.lucene.search.BooleanQuery;
+import org.apache.lucene.search.Query;
+import org.apache.lucene.search.TermQuery;
+import org.opensearch.neuralsearch.query.NeuralKNNQuery;
+import org.opensearch.neuralsearch.query.HybridQuery;
+
+import lombok.extern.log4j.Log4j2;
+
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * Registry for query text extractors that manages the extraction process
+ */
+@Log4j2
+public class QueryTextExtractorRegistry {
+
+    private final Map<Class<? extends Query>, QueryTextExtractor> extractors = new HashMap<>();
+
+    /**
+     * Creates a new registry with default extractors
+     */
+    public QueryTextExtractorRegistry() {
+        initialize();
+    }
+
+    /**
+     * Initializes the registry with default extractors
+     */
+    private void initialize() {
+        register(NeuralKNNQuery.class, new NeuralQueryTextExtractor());
+        register(TermQuery.class, new TermQueryTextExtractor());
+        register(HybridQuery.class, new HybridQueryTextExtractor(this));
+
+        // BooleanQueryTextExtractor needs a reference to this registry
+        // so we need to register it after creating the registry instance
+        register(BooleanQuery.class, new BooleanQueryTextExtractor(this));
+    }
+
+    /**
+     * Registers an extractor for a specific query type
+     *
+     * @param queryClass The query class to register for
+     * @param extractor The extractor to use for this query type
+     */
+    public <T extends Query> void register(Class<T> queryClass, QueryTextExtractor extractor) {
+        extractors.put(queryClass, extractor);
+    }
+
+    /**
+     * Extracts text from a query using the appropriate extractor
+     *
+     * @param query The query to extract text from
+     * @param fieldName The name of the field being highlighted
+     * @return The extracted query text
+     * @throws IllegalArgumentException if no suitable extractor is found or the extracted query text is empty
+     */
+    public String extractQueryText(Query query, String fieldName) {
+        // Find the most specific extractor for this query type
+        Class<?> queryClass = query.getClass();
+        QueryTextExtractor extractor;
+
+        // Look for an exact match first
+        extractor = extractors.get(queryClass);
+
+        // If no exact match, look for parent classes
```

Comment:
Wondering if there is such case where one query extend from another?

Reply:
We can remove this part since we covered common query cases, and if there is a case of null extractor found, then no highlight section will be returned, which is same behavior like existing highlighter.